cardano-crypto-tests/src/Test/Crypto/KES.hs

```diff
@@ -36,363 +64,741 @@ import Test.Crypto.Util (
+
+instance ( MonadSodium m
+         , MonadST m
+         , MEq m (SignKeyKES d)
+         , Eq (VerKeyKES d)
+         , KnownNat (SeedSizeKES d)
+         ) => MEq m (SignKeyKES (SumKES h d)) where
+  equalsM (SignKeySumKES s r v1 v2) (SignKeySumKES s' r' v1' v2') =
+    (s, r, PureMEq v1, PureMEq v2) ==! (s', r', PureMEq v1', PureMEq v2')
+
+deriving newtype instance (MEq m (SignKeyDSIGNM d)) => MEq m (SignKeyKES (CompactSingleKES d))
+
+instance ( MonadSodium m
+         , MonadST m
+         , MEq m (SignKeyKES d)
+         , Eq (VerKeyKES d)
+         , KnownNat (SeedSizeKES d)
+         ) => MEq m (SignKeyKES (CompactSumKES h d)) where
+  equalsM (SignKeyCompactSumKES s r v1 v2) (SignKeyCompactSumKES s' r' v1' v2') =
+    (s, r, PureMEq v1, PureMEq v2) ==! (s', r', PureMEq v1', PureMEq v2')
+
+testKESAlloc
+  :: forall v.
+     ( (forall m. (MonadSodium m, MonadThrow m, MonadST m) => KESSignAlgorithm m v)
+     , ContextKES v ~ ()
+     )
+  => Proxy v
+  -> String
+  -> TestTree
+testKESAlloc _p n =
+  testGroup n
+    [ testGroup "Forget mock"
+      [ testCase "genKey" $ testForgetGenKeyKES _p
+      , testCase "updateKey" $ testForgetUpdateKeyKES _p
+      ]
+    , testGroup "Low-level mlocked allocations"
+      [ testCase "genKey" $ testMLockGenKeyKES _p
+      -- , testCase "updateKey" $ testMLockUpdateKeyKES _p
+      ]
+    ]
+
+testForgetGenKeyKES
+  :: forall v.
+     ( KESSignAlgorithm (LogT (GenericEvent ForgetMockEvent) IO) v
+     )
+  => Proxy v
+  -> Assertion
+testForgetGenKeyKES _p = do
+  logVar <- newIORef []
+  let tracer :: Tracer (LogT (GenericEvent ForgetMockEvent) IO) (GenericEvent ForgetMockEvent)
+      tracer = Tracer (\ev -> liftIO $ modifyIORef logVar (++ [ev]))
+  runLogT tracer $ do
+    seed <- MLockedSeed <$> mlsbFromByteString (BS.replicate 1024 23)
+    sk <- genKeyKES @(LogT (GenericEvent ForgetMockEvent) IO) @(ForgetMockKES v) seed
+    mlockedSeedFinalize seed
+    forgetSignKeyKES sk
+  result <- readIORef logVar
+  assertEqual "number of log entries" 2 (length result)
+  assertBool "first entry is GEN" (isGEN . concreteEvent $ (result !! 0))
+  assertBool "second entry is DEL" (isDEL . concreteEvent $ (result !! 1))
+  return ()
+
+testForgetUpdateKeyKES
+  :: forall v.
+     ( KESSignAlgorithm (LogT (GenericEvent ForgetMockEvent) IO) v
+     , ContextKES v ~ ()
+     )
+  => Proxy v
+  -> Assertion
+testForgetUpdateKeyKES _p = do
+  logVar <- newIORef []
+  let tracer :: Tracer (LogT (GenericEvent ForgetMockEvent) IO) (GenericEvent ForgetMockEvent)
+      tracer = Tracer (\ev -> liftIO $ modifyIORef logVar (++ [ev]))
+  runLogT tracer $ do
+    seed <- MLockedSeed <$> NaCl.mlsbFromByteString (BS.replicate 1024 23)
+    sk <- genKeyKES @(LogT (GenericEvent ForgetMockEvent) IO) @(ForgetMockKES v) seed
+    mlockedSeedFinalize seed
+    msk' <- updateKES () sk 0
+    case msk' of
+      Just sk' -> forgetSignKeyKES sk'
+      Nothing -> return ()
+  threadDelay 1000000
```

Comment:
Is this delay here to wait for GC? If that's the case it would be better to add `performGC` and reduce the delay to half a second or something.

Reply:
It was, but it doesn't actually sense anymore, because the ForgetMock logging we have now does not infect finalizers, and so this test cannot check whether the finalizer runs and correctly forgets the mlocked memory; and besides, the value of these tests is that they verify that the KES implementations recurse correctly, not that they use the underlying mlocked memory correctly - The AllocLog-based tests do this though, so I'm now rewriting these tests to reflect proper usage, and to check the entire ForgetMock log.

Reply:
In fact, this has me thinking whether the entire ForgetMock-based testing is useful at all; as it stands these tests really just test that the functions we call actually get called. It could be more useful when testing larger chunks of code that perform multiple KES operations though, so I'll leave it in.